### PR TITLE
Whitelist plugin dlls

### PR DESCRIPTION
**Pull request by AmbulantRex**
**Changes**
* Allows a plugin to pre-package a `meta.json` file which can define a list of assemblies to selectively whitelist. If not provided, defaults to current behavior of loading all of a plugin's `*.dll` files. 

**Issues**
* https://github.com/jellyfin/jellyfin/discussions/9523

### 🌎 Context
Plugins that leverage wrapper libraries for .NET native interoperability may have unmanaged DLLs in the plugin directory. Examples of this would include things like machine learning libraries that are commonly wrappers around popular native libraries (see https://github.com/charlesw/tesseract as an example). This change allows those plugins to selectively ignore some DLLs to load themselves later. 

### 🔒 Security Note
The list of DLLs for the plugin is drawn from the plugin's directory by searching for `*.dll` files. It uses the `Directory.GetFiles()` method to do so guaranteeing those files exist within that folder. Allowing plugin developers to list relative paths to DLLs could easily expose a [path traversal risk](https://owasp.org/www-community/attacks/Path_Traversal). 

This change attempts to use the assembly list from the manifest for comparison only and still leverages the same directory listing method above. It will then only load DLLs that are in the path AND whitelisted. This is an attempt at safelisting that aims to prevent things like `../` traversal or `~/` expansion from allowing a plugin developer to load assemblies outside of the plugin folder. Even still, there may be vulnerabilities with this that I am unaware of and should be considered carefully during review. 

**_It should be noted that loading assemblies directly from untrusted sources is highly discouraged in any case._**

### CODE REVIEW

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:768]:
```suggestion
            ArgumentNullException.ThrowIfNull(plugin);
```

File: MediaBrowser.Common/Plugins/PluginManifest.cs
Comment (crobibero) [inline MediaBrowser.Common/Plugins/PluginManifest.cs:115]:
```suggestion
        public IReadOnlyList<string> Assemblies { get; set; }
```

File: MediaBrowser.Model/Updates/VersionInfo.cs
Comment (crobibero) [inline MediaBrowser.Model/Updates/VersionInfo.cs:83]:
```suggestion
        public IReadOnlyList<string> Assemblies { get; set; } = Array.Empty<string>();
```

File: tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json
Comment (crobibero) [inline tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json:28]:
```suggestion
        ]
```

File: MediaBrowser.Common/Plugins/PluginManifest.cs
Comment (crobibero) [inline MediaBrowser.Common/Plugins/PluginManifest.cs:27]:
```suggestion
            Assemblies = Array.Empty<string>();
```

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:108]:
```suggestion
            var newSubPathTrimmed = newSubPath.AsSpan().TrimEnd(newDirectorySeparatorChar);
```

This change isn't required for me

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:131]:
```suggestion
        [return: NotNullIfNotNull(nameof(path))]
        public static string? NormalizePath(this string? path)
```

then remove `!` from L123 (and anywhere else)

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:162]:
```suggestion
            return path.NormalizePath(newSeparator);
```

`path` is known to not be null here

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:187]:
```suggestion
            return newSeparator == Bs ? path.Replace(Fs, newSeparator) : path.Replace(Bs, newSeparator);
```

`path` is known to not be null here

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:172]:
```suggestion
        [return: NotNullIfNotNull(nameof(path))]
        public static string? NormalizePath(this string? path, char newSeparator)
```

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:142]:
```suggestion
        [return: NotNullIfNotNull(nameof(path))]
        public static string? NormalizePath(this string? path, out char separator)
```

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:90]:
```suggestion
            subPath = subPath.NormalizePath(out var newDirectorySeparatorChar);
```

File: Emby.Server.Implementations/Library/PathExtensions.cs
Comment (crobibero) [inline Emby.Server.Implementations/Library/PathExtensions.cs:91]:
```suggestion
            path = path.NormalizePath(newDirectorySeparatorChar);
```

**Review by crobibero — APPROVED** [review]

File: tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json
Comment (oddstr13) [inline tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json:17]: Why would the assemblies information need to be present in the repo manifest?
Would it not be more appropriate to only include it in the plugin metadata (`meta.json`)?

The content of the items in the versions array and the `meta.json` file in the plugin archive are not exact duplicates.

```json
{
    "category": "Metadata",
    "changelog": "### Bug Fixes ###\n- fix client string (#19) @Shadowghost\n\n### Code or Repo Maintenance ###\n- Fix readme (#20) @Shadowghost",
    "description": "This plugin provides images from the Cover Art Archive https://musicbrainz.org/doc/Cover_Art_Archive and depends on the MusicBrainz metadata provider to know what images belong where\n",
    "guid": "8119f3c6-cfc2-4d9c-a0ba-028f1d93e526",
    "imageUrl": "https://repo.jellyfin.org/releases/plugin/images/jellyfin-plugin-coverartarchive.png",
    "name": "Cover Art Archive",
    "overview": "MusicBrainz Cover Art Archive",
    "owner": "jellyfin",
    "targetAbi": "10.8.0.0",
    "timestamp": "2022-01-08T17:22:30Z",
    "version": "7.0.0.0"
}
```

Further I'd like to note that the Anime plugin is deprecated, and as such might not be the best test target (it was split into separate plugins)

Comment (oddstr13) [inline tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json:17]: Why would the assemblies information need to be present in the repo manifest?
Would it not be more appropriate to only include it in the plugin metadata (`meta.json`)?

The content of the items in the versions array and the `meta.json` file in the plugin archive are not exact duplicates.

```json
{
    "category": "Metadata",
    "changelog": "### Bug Fixes ###\n- fix client string (#19) @Shadowghost\n\n### Code or Repo Maintenance ###\n- Fix readme (#20) @Shadowghost",
    "description": "This plugin provides images from the Cover Art Archive https://musicbrainz.org/doc/Cover_Art_Archive and depends on the MusicBrainz metadata provider to know what images belong where\n",
    "guid": "8119f3c6-cfc2-4d9c-a0ba-028f1d93e526",
    "imageUrl": "https://repo.jellyfin.org/releases/plugin/images/jellyfin-plugin-coverartarchive.png",
    "name": "Cover Art Archive",
    "overview": "MusicBrainz Cover Art Archive",
    "owner": "jellyfin",
    "targetAbi": "10.8.0.0",
    "timestamp": "2022-01-08T17:22:30Z",
    "version": "7.0.0.0"
}
```

Further I'd like to note that the Anime plugin is deprecated, and as such might not be the best test target (it was split into separate plugins, and is no longer maintained)

**Review by oddstr13 — CHANGES_REQUESTED** [review]

File: tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json
Comment (AmbulantRex) [reply tests/Jellyfin.Server.Implementations.Tests/Test Data/Updates/manifest-stable.json:17]: The manifest is the repository that these are drawn from. The `meta.json` is ultimately derived from the version information supplied by the `manifest.json` seen here: 
https://github.com/jellyfin/jellyfin/blob/bb5bf0277ae33fd4668e00bd2f1a85e89aef466a/Emby.Server.Implementations/Plugins/PluginManager.cs#L413-L428

I don't go out of my way to change the current behavior between the `manifest.json` and `meta.json` in the case that the `meta.json` already exists. Since my manual testing involved supplying the `meta.json` directly, it seems like it would work in either case. 

The use of the Anime plugin was merely out of convenience. The `manifest-stable.json` is the test data response used in other testing. I wanted to have a blend of packages _with_ assemblies and those _without_. If this still feels inappropriate, I can re-assess.

**Comment by AmbulantRex** [comment]
@oddstr13 did my changes resolve your concerns?

**Review by oddstr13 — APPROVED** [review]

**Comment by oddstr13** [comment]
GitHub wouldn't allow me to just dismiss my change request, so, please do a proper review before merging @jellyfin/server

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:435]: Double combine?

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:483]: This should be moved to `static readonly` at the class level.

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:485]: I know this is more "future-proof" but I am not a fan of iterating through all the properties. This should be specific in the properties it maps

File: MediaBrowser.Model/Updates/VersionInfo.cs
Comment (crobibero) [inline MediaBrowser.Model/Updates/VersionInfo.cs:2]: These don't seem to be needed

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (AmbulantRex) [reply Emby.Server.Implementations/Plugins/PluginManager.cs:485]: Perhaps I could just move the reflection aspect into the tests so that new unmapped properties are detected early.  I'll take a look.

**Comment by AmbulantRex** [comment]
@crobibero I've adjusted the logic to do explicit mapping now. Some of the reconciliation was a bit leaky so I also consolidated it under a new method. My tests now use the `Assert.Equivalent` when checking the `PluginManifest`. This will hopefully help catch issues with future un-mapped properties.

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:486]:
```suggestion
                manifest.Timestamp = localManifest.Timestamp.Equals(default) ? manifest.Timestamp : localManifest.Timestamp;
```

Then we don't need the `TypeExtension`

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:463]:
```suggestion
                using var metaStream = File.OpenRead(metafile);
                var localManifest = await JsonSerializer.DeserializeAsync<PluginManifest>(metaStream, _jsonOptions).ConfigureAwait(false);
                localManifest ??= new PluginManifest();
```

And make this function async

File: Emby.Server.Implementations/Plugins/PluginManager.cs
Comment (crobibero) [inline Emby.Server.Implementations/Plugins/PluginManager.cs:845]:
```suggestion
                    if (!canonicalized.StartsWith(plugin.Path.NormalizePath(), StringComparison.Ordinal))
```

**Comment by AmbulantRex** [comment]
@crobibero have my latest changes resolved your concerns?

**Review by crobibero — APPROVED** [review]

**Comment by AmbulantRex** [comment]
@crobibero, is there anything more to do before this can be merged?

**Comment by crobibero** [comment]
Another server team member will need to review the code changes.

**Comment by AmbulantRex** [comment]
@Bond-009 perhaps you could give this a look as well?

**Comment by AmbulantRex** [comment]
@crobibero sorry to be a pest, but is there a good way to tell who is on the server team? Is there someone I can summon?

**Comment by crobibero** [comment]
Please stop pinging people. It will get reviewed when someone has time

**Pull request merged**
